### ISSUE TITLE
Setting "Mixed" for property

### ISSUE THREAD

**Issue by wasya243** [typegoose/typegoose#882]
Hello, I have a question regarding Setting "Mixed" for property. I already seen this question https://github.com/typegoose/typegoose/issues/239, but not sure I got the answers I needed. Take a look at the code

```ts
import mongoose from 'mongoose';
import {
  prop,
  getModelForClass,
} from '@typegoose/typegoose';

export class Settings {
  @prop({ required: true })
  enabled!: boolean;
}

export class User {
  @prop({required: true})
  firstName!: string;

  @prop({required: true})
  lastName!: string;

  @prop({ type: () => Map<string, any> })
  features?: Map<string, any>;

  @prop({ type: () => Settings, required: true })
  settings!: Settings;
}

export const UserModel = getModelForClass(User);

async function main() {
  try {
    await mongoose.connect('mongodb://localhost:27017/', {dbName: 'test-mixed-typegoose'});

    await UserModel.create({
      firstName: 'Vasyl',
      lastName: 'Kh',
      settings: {
        enabled: true
      }
    })

    process.exit(0);
  } catch (err) {
    console.log(err);
    process.exit(1);
  }
}

main();
```

When I run it I get Setting "Mixed" for property "User.features". I know how to disable it, but does not typegoose/mongoose recognise Map<string, any> as a type instead of setting it into Mixed? Why it does not complain about **settings** property (isn't it because it treats it like a Schema type?)?

**Comment by wasya243** [comment]
Wait a sec, https://mongoosejs.com/docs/schematypes.html#maps -> is a type, so why I should change it into Mixed?

**Comment by wasya243** [comment]
@hasezoey or maybe it's the answer I need?
```
  @prop({ type: () => new Passthrough({ type: Map, of: String }, true)  })
  public features?: Map<string, string>;
```

**Comment by hasezoey** [comment]
> or maybe it's the answer I need? https://typegoose.github.io/typegoose/docs/api/decorators/prop/#passthrough-class

no this is not the answer for this case

> Basically I need somehow to transform this (arbitrary object -> logically it's a Map of string-any) into typegoose format

that is what `Mixed` is for: arbitrary data

> And yeap, I have no runtime error.

very weird, because `type: () => Map<string, any>` is not valid js syntax, is this maybe something in the example but not in your given code?

---

typegoose works by automatically inferring the `PropType` via `emitDecoratorMetadata` if available, this can be set manually as the second parameter to `@prop` (like `@prop({ someoption: true }, PropType.ARRAY)`, see [`@prop`](https://typegoose.github.io/typegoose/docs/api/decorators/prop) definition for usage and parameters and [Use Without "emitDecoratorMetadata"](https://typegoose.github.io/typegoose/docs/guides/use-without-emitDecoratorMetadata) for manually setting `PropType` or for usage in a environment which does not support that option.

downside to `emitDecoratorMetadata` is that it only outputs the basic type, which is enough for simple types like `String`, but is not enough and just outputs `Array` or `Map`, without any option to get what the key / value type is meant to be.

in old versions of typegoose you had to use the exact same options as mongoose (like `type: Map, of: String`), which made it rather redundant information, so newer versions of typegoose (at least since 7.0) changed to use a simpler system of only having one `type` option which always refers to the actual data type, and setting the "wrapper" type as a second argument, if necessary.

TL;DR: for usage / examples see the [`@prop PropType.MAP`](https://typegoose.github.io/typegoose/docs/api/decorators/prop#map-options)

**Comment by wasya243** [comment]
@hasezoey hello, so as far as I understood I should use the following in order to emulate arbitrary key-value object?

```
  @prop({ type: () => String }, PropType.MAP)
  public features?: Map<string, string>;

  @prop({ type: () => Object }, PropType.MAP)
  public features2?: Map<string, Object>;
```

I've tested it, seems to be ok. Returns Map when I reference these fields on queried object.

Can you please add a ref then, 'cause not sure that I follow how **type: () =>** thing works. I though this is where you should specify type (btw what type? how it's used and for what?)

**Comment by hasezoey** [comment]
>  so as far as I understood I should use the following in order to emulate arbitrary key-value object?

for `features2`, yes, though likely you will also have to set `allowMixed: Severity.ALLOW` for that property (because typegoose cant differentiate between implicit and explicit Mixed types)

> Can you please explain which type is used for determining how to store data in mongodb and which type is used to convert data from db to smth like ts Map?

didnt i do this in my [last comment](https://github.com/typegoose/typegoose/issues/882#issuecomment-1728242579)?

both types are the same, it is just that typegooes prop option `type` will always refer to the *inner* type, the *wrapper* type is defined via the second argument to `@prop`, which can be automatically inferred if `emitDecoratorMetadata` is enabled.

one special case being `Map<String, String[]>`, which does not need setting some kind of `PropType` for the inner arrays:

```ts
@prop({ type: () => [String] })
property: Map<String, String[]>
```

(though unless the passthrough class is used you cannot create something like `[Map<String, String>]` (a array of maps) or a map of maps without using a intermediate class)

TL;DR: mongoose to typegoose 1:1 explicit definition example:

```ts
const schema = new mongoose.Schema({
  property: {
    type: Map,
    of: {
      type: String,
    },
  },
});

class Schema {
  @prop({ type: () => String }, PropType.MAP)
  public property?: Map<string, string>;
}
```

PS: maybe looking at the actual code typegoose uses for Maps (with primitives) will help instead of me trying to just explain it 

https://github.com/typegoose/typegoose/blob/861ea1a94e90a6de757e85359903f24777ed5d94/src/internal/processProp.ts#L395-L415

**Comment by wasya243** [comment]
@hasezoey okey, I understood. So I did a wrong thing when writing this?
```
        @prop({ type: () => [Number] })
      lsList?: number[];

              @prop({ type: () => [String] })
      ddList?: string[];

          @prop({ type: () => [HistoryRecord] })
      history?: HistoryRecord[];

          @prop({ type: () => [mongoose.Schema.Types.Mixed] })
      history_transactions?: mongoose.Schema.Types.Mixed[];
```
Could you please translate it into right format, so that I solidified this knowledge.

**Comment by hasezoey** [comment]
the code (mentioned in https://github.com/typegoose/typegoose/issues/882#issuecomment-1730905746) looks correct to me

just note that instead of mixed you *could* also use other types, like `Object` or `Record<something, something>` (typescript will emit `Object` for that type)

also unless you have it disabled anywhere more globally, you likely need to add `allowMixed: Severity.ALLOW` to the `history_transactions` prop options (to not get the mixed warning)

**Comment by hasezoey** [comment]
i dont know what you are trying to achieve here. is `objSchema` meant as a "allow all key-value things"? when yes, i think `Record<string, unknown>` would be the more correct type instead of `Map<string, unknown>` if you want a 1:1 conversion (though to my knowledge schemas without `strict: false` dont allow any other value than defined in the schema)

**Comment by wasya243** [comment]
@hasezoey yeap, it's more like key-value, where value -> can be anything. So, I should use Record<string, unknown> instead?

**Comment by wasya243** [comment]
@hasezoey okey, thanks
